City-of-Helsinki/helsinki-paatokset
# UHF-8855: Fetch composition and roles from policymaker data

# [UHF-8855](https://helsinkisolutionoffice.atlassian.net/browse/UHF-8855)

## What was done
* Added new migration for fetching policymaker composition from a new endpoint
* Added a new proxy endpoint for decisionmaker data (slightly different from the organization endpoint)
* Added composition JSON-field to policymaker nodes
* Fetch organization composition from policymaker nodes instead of meetings. This affects:
  * Showing the composition on the policymaker page
  * Showing the person's roles on a trustee page
  * Fetching city council members for the Datapumppu integration

## How to install (with VPN connection)
* This requires a VPN connection and authentication with AHJO API, because the new proxy URL doesn't exist yet in production or staging. More info: https://helsinkisolutionoffice.atlassian.net/wiki/spaces/PP/pages/7274659859/OpenID+Connect+ja+lokaali+proxy
* Comment out the following values in `.env.local`:
  * `AHJO_PROXY_BASE_URL`, `SKIP_AUTH_HEADERS`
  * Run `make up` to make sure the environment variables are updated
  * Run `printenv` to check. Also make sure `DRUPAL_REVERSE_PROXY_ADDRESS` is `helsinki-paatokset.docker.so` 
* Add this value to your `local.settings.php` temporarily so you can connect to local aggregation URLS: `$settings['http_client_config']['verify'] = FALSE;`  
* Make sure your instance is up and running on correct branch.
  * `git checkout UHF-8855-update-roles-and-memberships`
  * `make fresh`
* Run `make drush-cr`
* Authenticate with AHJO API:
  * Login to VPN
  * Go to https://helsinki-paatokset.docker.so/fi/admin/ahjo-open-id/settings and change the "Internal callback URL" setting to https://helsinki-paatokset.docker.so/ahjo-api/login
  * Follow the authentication flow on https://helsinki-paatokset.docker.so/fi/admin/ahjo-open-id (ask Juho for the credentials)
  * If everything works, the system should fetch the access token correctly
  * Verify that the connection works by opening this URL: https://helsinki-paatokset.docker.so/fi/ahjo-proxy/decisionmaker/single/00400
    * If the dataset is empty, the connection is not working. 

## How to install (without VPN connection)
* Make sure you have the following values in `.env.local`:
  * `AHJO_PROXY_BASE_URL=https://drupal-paatokset.stage.hel.ninja/`,
  * `LOCAL_PROXY_API_KEY=[API key for staging]`
  * `SKIP_AUTH_HEADERS=1`
* Run `make up` to make sure the environment variables are updated
  * Run `printenv` to check. Also make sure `DRUPAL_REVERSE_PROXY_ADDRESS` is `helsinki-paatokset.docker.so` 
* Make sure your instance is up and running on correct branch.
  * `git checkout UHF-8855-update-roles-and-memberships`
  * `make fresh`

## How to test
* Update decisionmaker and trustee nodes: `drush mim ahjo_decisionmakers:all --update;drush mim ahjo_decisionmakers:all_sv --update` and `drush migrate-import ahjo_trustees:all --update;drush migrate-import ahjo_trustees:all_sv --update`
* [x] Check that the composition and roles lists are empty for now, since the data hasn't been fetched yet:
  * The city council member list should be empty: https://helsinki-paatokset.docker.so/fi/paattajat
  * The members filter list should be missing: https://helsinki-paatokset.docker.so/fi/paattajat/kaupunginvaltuusto
  * The roles list should only have chairmanships, which are fetched from trustee data: https://helsinki-paatokset.docker.so/fi/paattajat/arhinpa
  * Also check a few more trustee nodes
* [x] The datapumppu integration should fail because it doesn't find any council member nodes: `drush migrate:import datapumppu_statements -v` 
* [x] Run the integration: `drush mim ahjo_org_composition -v`. This should update around 52 nodes.
* [x] Run the same command again. Since nothing was changed it shouldn't update anything.
* [x] Edit the migration YML file `public/modules/custom/paatokset_ahjo_api/migrations/ahjo_org_composition.yml` and change the source plugin settings:
  * `orgs:active` -> `orgs: all`
* [x] Rerun the command: `drush cr;drush mim ahjo_org_composition -v`. This should update about 40 more nodes.
* Try running the drush commands. Verify this by first manually editing a few policymaker node composition (for example remove a few entries). After running the command, the changes should be reverted because these run the migrations with the force update flag. Unfortunately the script doesn't pass the verbose flag to the migration, so we can't directly see how many nodes it updated.
  * [x] `drush org-composition:fetch-all -v` 
  * [x] `drush org-composition:fetch-by-id --ids=00400,02900-v` (edit the city council and board nodes before running this).
* [x] Run the datapumppu integration, which should work now: `drush migrate:import datapumppu_statements -v`
* Check the listings again:
  * [x] The city council member list should work now: https://helsinki-paatokset.docker.so/fi/paattajat
  * [x] The members filter list should work: https://helsinki-paatokset.docker.so/fi/paattajat/kaupunginvaltuusto
  * [x] Also check a few other organizations to make sure their members lists work
  * [x] Trustee memberships and roles should also work: https://helsinki-paatokset.docker.so/fi/paattajat/arhinpa
    * [x] The roles and org names should be translated to swedish also
  * [x] Also check a few other trustees (might be easier to open the profile pages directly from the organization member lists)
* Check that cache invalidation works (make sure you have Drupal's caching turned on)
  * Open any trustee page with some organization memberships, then find the policymaker node
  * [x] Edit the policymaker node and remove the persons entry from the composition JSON field
  * [x] Reload the trustee node (without clearing caches). The membership should be removed now
  * [x] Run `drush oc:id --ids=[policymaker ID]`. Check that the entry you deleted is back now.
  * [x] Reload the trustee node again. The membership should be back
  * [x] Delete the policymaker node, then reload the trustee node. The membership should be missing
* [ ] Is the current `ahjo_org_composition` migration definition enough to catch updated compositions, or does it need some additions?
* [ ] Can you think of situations where having two different migrations touch the same nodes would cause issues?
* [x] Check that updated documentation is OK:
  * https://helsinkisolutionoffice.atlassian.net/wiki/spaces/PP/pages/8333819923/Kriittiset+toiminnallisuudet+ja+ongelmatilanteiden+selvitys 
  * https://helsinkisolutionoffice.atlassian.net/wiki/spaces/PP/pages/7839711401/AHJO+API+migraatiot+prosessointijonot+ja+datan+aggregointi
  * https://helsinkisolutionoffice.atlassian.net/wiki/spaces/PP/pages/7839383830/Migraatiokomentojen+ajastus
  * https://helsinkisolutionoffice.atlassian.net/wiki/spaces/PP/pages/7874904065/Sis+iset+rajapinnat+ja+proxy+Ahjon+suuntaan
  * https://helsinkisolutionoffice.atlassian.net/wiki/spaces/PP/pages/7839973578/Integraation+sis+lt+tyypit+ja+sis+ll+n+rikastaminen

## Continuous documentation

[UHF-8855]: https://helsinkisolutionoffice.atlassian.net/browse/UHF-8855?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
